streams/src/main/java/org/apache/kafka/streams/state/internals/Segments.java

```diff
@@ -77,10 +77,10 @@ Segment getOrCreateSegment(final long segmentId, final ProcessorContext context)
             if (!isSegment(segment, segmentId)) {
                 cleanup(segmentId);
             }
-            if (!segments.containsKey(key)) {
-                Segment newSegment = new Segment(segmentName(segmentId), name, segmentId);
+            Segment newSegment = new Segment(segmentName(segmentId), name, segmentId);
+            Segment previousSegment = segments.putIfAbsent(key, newSegment);
+            if (previousSegment == null) {
```

Comment:
Thanks, looks good.

Yes, it's O(1), but a lot less efficient than returning a local variable. Take a look. :)

```java
public V get(Object key) {
        Node<K,V>[] tab; Node<K,V> e, p; int n, eh; K ek;
        int h = spread(key.hashCode());
        if ((tab = table) != null && (n = tab.length) > 0 &&
            (e = tabAt(tab, (n - 1) & h)) != null) {
            if ((eh = e.hash) == h) {
                if ((ek = e.key) == key || (ek != null && key.equals(ek)))
                    return e.val;
            }
            else if (eh < 0)
                return (p = e.find(h, key)) != null ? p.val : null;
            while ((e = e.next) != null) {
                if (e.hash == h &&
                    ((ek = e.key) == key || (ek != null && key.equals(ek))))
                    return e.val;
            }
        }
        return null;
    }
```